File: apps/api/drift_api/repositories.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, Optional, Sequence


@dataclass(frozen=True)
class UploadRecord:
    upload_id: str
    source_archive_uri: str
    document_uris: Sequence[str]
    project_name: Optional[str] = None
    source_file_name: Optional[str] = None
    document_file_names: Sequence[str] = field(default_factory=tuple)


@dataclass(frozen=True)
class JobRecord:
    job_id: str
    status: str
    source_archive_uri: str
    document_uris: Sequence[str]
    results_prefix: str
    upload_id: Optional[str] = None
    project_name: Optional[str] = None
    artifact_paths: Dict[str, str] = field(default_factory=dict)
    error_message: Optional[str] = None


class FirestoreUploadRepository:
    def __init__(self, collection_name: str, firestore_client=None) -> None:
        if firestore_client is None:
            from google.cloud import firestore

            firestore_client = firestore.Client()
            self._server_timestamp = firestore.SERVER_TIMESTAMP
        else:
            self._server_timestamp = getattr(firestore_client, "SERVER_TIMESTAMP", None)
        self._collection = firestore_client.collection(collection_name)

    def create(self, record: UploadRecord) -> None:
        self._collection.document(record.upload_id).set(
            {
                "upload_id": record.upload_id,
                "source_archive_uri": record.source_archive_uri,
                "document_uris": list(record.document_uris),
                "project_name": record.project_name,
                "source_file_name": record.source_file_name,
                "document_file_names": list(record.document_file_names),
                "created_at": self._server_timestamp,
            }
        )

    def get(self, upload_id: str) -> Optional[UploadRecord]:
        snapshot = self._collection.document(upload_id).get()
        if not snapshot.exists:
            return None
        data = snapshot.to_dict() or {}
        return UploadRecord(
            upload_id=upload_id,
            source_archive_uri=str(data.get("source_archive_uri") or data.get("sourceArchiveUri") or ""),
            document_uris=tuple(data.get("document_uris") or data.get("documentUris") or ()),
            project_name=data.get("project_name") or data.get("projectName"),
            source_file_name=data.get("source_file_name") or data.get("sourceFileName"),
            document_file_names=tuple(
                data.get("document_file_names") or data.get("documentFileNames") or ()
            ),
        )


class FirestoreJobRepository:
    def __init__(self, collection_name: str, firestore_client=None) -> None:
        if firestore_client is None:
            from google.cloud import firestore

            firestore_client = firestore.Client()
            self._server_timestamp = firestore.SERVER_TIMESTAMP
        else:
            self._server_timestamp = getattr(firestore_client, "SERVER_TIMESTAMP", None)
        self._collection = firestore_client.collection(collection_name)

    def create_queued(self, record: JobRecord) -> None:
        self._collection.document(record.job_id).set(
            {
                "job_id": record.job_id,
                "upload_id": record.upload_id,
                "project_name": record.project_name,
                "source_archive_uri": record.source_archive_uri,
                "document_uris": list(record.document_uris),
                "results_prefix": record.results_prefix,
                "status": "queued",
                "artifact_paths": {},
                "error_message": None,
                "created_at": self._server_timestamp,
                "updated_at": self._server_timestamp,
            }
        )

    def mark_failed(self, job_id: str, error_message: str) -> None:
        self._collection.document(job_id).set(
            {
                "status": "failed",
                "error_message": error_message,
                "updated_at": self._server_timestamp,
            },
            merge=True,
        )

    def get(self, job_id: str) -> Optional[JobRecord]:
        snapshot = self._collection.document(job_id).get()
        if not snapshot.exists:
            return None
        data = snapshot.to_dict() or {}
        return JobRecord(
            job_id=job_id,
            status=str(data.get("status", "")),
            upload_id=data.get("upload_id") or data.get("uploadId"),
            project_name=data.get("project_name") or data.get("projectName"),
            source_archive_uri=str(data.get("source_archive_uri") or data.get("sourceArchiveUri") or ""),
            document_uris=tuple(data.get("document_uris") or data.get("documentUris") or ()),
            results_prefix=str(data.get("results_prefix") or data.get("resultsPrefix") or ""),
            artifact_paths=dict(data.get("artifact_paths") or data.get("artifactPaths") or {}),
            error_message=data.get("error_message") or data.get("errorMessage"),
        )

File: apps/api/src/repositories.ts
import { FieldValue, Firestore } from "@google-cloud/firestore";

type FirestoreCollection = {
  doc(id: string): {
    set(data: Record<string, unknown>, options?: { merge?: boolean }): Promise<void>;
    get(): Promise<{ exists: boolean; data(): Record<string, unknown> | undefined }>;
  };
};

type FirestoreClient = {
  SERVER_TIMESTAMP?: unknown | (() => unknown);
  collection(name: string): FirestoreCollection;
};

export type UploadRecord = Readonly<{
  uploadId: string;
  sourceArchiveUri: string;
  documentUris: readonly string[];
  projectName: string | null;
  sourceFileName: string | null;
  documentFileNames: readonly string[];
}>;

export type JobRecord = Readonly<{
  jobId: string;
  status: string;
  sourceArchiveUri: string;
  documentUris: readonly string[];
  resultsPrefix: string;
  uploadId: string | null;
  projectName: string | null;
  artifactPaths: Record<string, string>;
  errorMessage: string | null;
}>;

export interface UploadRepository {
  create(record: UploadRecord): Promise<void>;
  get(uploadId: string): Promise<UploadRecord | null>;
}

export interface JobRepository {
  createQueued(record: JobRecord): Promise<void>;
  markFailed(jobId: string, errorMessage: string): Promise<void>;
  get(jobId: string): Promise<JobRecord | null>;
}

export class FirestoreUploadRepository implements UploadRepository {
  private readonly serverTimestamp: unknown;
  private readonly collection: FirestoreCollection;

  constructor(collectionName: string, firestoreClient: FirestoreClient | null = null) {
    const client = firestoreClient ?? (new Firestore() as unknown as FirestoreClient);
    this.serverTimestamp = firestoreClient ? serverTimestampFromClient(firestoreClient) : FieldValue.serverTimestamp();
    this.collection = client.collection(collectionName);
  }

  async create(record: UploadRecord): Promise<void> {
    await this.collection.doc(record.uploadId).set({
      upload_id: record.uploadId,
      source_archive_uri: record.sourceArchiveUri,
      document_uris: [...record.documentUris],
      project_name: record.projectName,
      source_file_name: record.sourceFileName,
      document_file_names: [...record.documentFileNames],
      created_at: this.serverTimestamp,
    });
  }

  async get(uploadId: string): Promise<UploadRecord | null> {
    const snapshot = await this.collection.doc(uploadId).get();
    if (!snapshot.exists) {
      return null;
    }
    const data = snapshot.data() || {};
    return uploadRecord({
      uploadId,
      sourceArchiveUri: String(data.source_archive_uri || data.sourceArchiveUri || ""),
      documentUris: stringArray(data.document_uris || data.documentUris),
      projectName: optionalString(data.project_name || data.projectName),
      sourceFileName: optionalString(data.source_file_name || data.sourceFileName),
      documentFileNames: stringArray(data.document_file_names || data.documentFileNames),
    });
  }
}

export class FirestoreJobRepository implements JobRepository {
  private readonly serverTimestamp: unknown;
  private readonly collection: FirestoreCollection;

  constructor(collectionName: string, firestoreClient: FirestoreClient | null = null) {
    const client = firestoreClient ?? (new Firestore() as unknown as FirestoreClient);
    this.serverTimestamp = firestoreClient ? serverTimestampFromClient(firestoreClient) : FieldValue.serverTimestamp();
    this.collection = client.collection(collectionName);
  }

  async createQueued(record: JobRecord): Promise<void> {
    await this.collection.doc(record.jobId).set({
      job_id: record.jobId,
      upload_id: record.uploadId,
      project_name: record.projectName,
      source_archive_uri: record.sourceArchiveUri,
      document_uris: [...record.documentUris],
      results_prefix: record.resultsPrefix,
      status: "queued",
      artifact_paths: {},
      error_message: null,
      created_at: this.serverTimestamp,
      updated_at: this.serverTimestamp,
    });
  }

  async markFailed(jobId: string, errorMessage: string): Promise<void> {
    await this.collection.doc(jobId).set(
      {
        status: "failed",
        error_message: errorMessage,
        updated_at: this.serverTimestamp,
      },
      { merge: true },
    );
  }

  async get(jobId: string): Promise<JobRecord | null> {
    const snapshot = await this.collection.doc(jobId).get();
    if (!snapshot.exists) {
      return null;
    }
    const data = snapshot.data() || {};
    return jobRecord({
      jobId,
      status: String(data.status || ""),
      uploadId: optionalString(data.upload_id || data.uploadId),
      projectName: optionalString(data.project_name || data.projectName),
      sourceArchiveUri: String(data.source_archive_uri || data.sourceArchiveUri || ""),
      documentUris: stringArray(data.document_uris || data.documentUris),
      resultsPrefix: String(data.results_prefix || data.resultsPrefix || ""),
      artifactPaths: stringRecord(data.artifact_paths || data.artifactPaths),
      errorMessage: optionalString(data.error_message || data.errorMessage),
    });
  }
}

export function uploadRecord({
  uploadId,
  sourceArchiveUri,
  documentUris,
  projectName = null,
  sourceFileName = null,
  documentFileNames = [],
}: {
  uploadId: string;
  sourceArchiveUri: string;
  documentUris: readonly string[];
  projectName?: string | null;
  sourceFileName?: string | null;
  documentFileNames?: readonly string[];
}): UploadRecord {
  return {
    uploadId,
    sourceArchiveUri,
    documentUris: Object.freeze([...documentUris]),
    projectName,
    sourceFileName,
    documentFileNames: Object.freeze([...documentFileNames]),
  };
}

export function jobRecord({
  jobId,
  status,
  sourceArchiveUri,
  documentUris,
  resultsPrefix,
  uploadId = null,
  projectName = null,
  artifactPaths = {},
  errorMessage = null,
}: {
  jobId: string;
  status: string;
  sourceArchiveUri: string;
  documentUris: readonly string[];
  resultsPrefix: string;
  uploadId?: string | null;
  projectName?: string | null;
  artifactPaths?: Record<string, string>;
  errorMessage?: string | null;
}): JobRecord {
  return {
    jobId,
    status,
    sourceArchiveUri,
    documentUris: Object.freeze([...documentUris]),
    resultsPrefix,
    uploadId,
    projectName,
    artifactPaths: { ...artifactPaths },
    errorMessage,
  };
}

function serverTimestampFromClient(client: FirestoreClient): unknown {
  if (typeof client.SERVER_TIMESTAMP === "function") {
    return client.SERVER_TIMESTAMP();
  }
  return client.SERVER_TIMESTAMP ?? null;
}

function optionalString(value: unknown): string | null {
  return value == null ? null : String(value);
}

function stringArray(value: unknown): string[] {
  return Array.isArray(value) ? value.map((item) => String(item)) : [];
}

function stringRecord(value: unknown): Record<string, string> {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return {};
  }
  return Object.fromEntries(Object.entries(value).map(([key, item]) => [key, String(item)]));
}
